File: dp/kadanes/max_sum_subarray.cpp
#include<iostream>
#include<vector>
using namespace std;

int maxSumSubarray(vector<int> array);

int maxSumSubarray(vector<int> array){
    int maxSoFar = array[0];
    int maxEndHere = array[0];
    for(int i=1;i<array.size();i++){
        int num = array[i];
        maxEndHere = max(num,  maxEndHere+num);
        maxSoFar = max(maxEndHere, maxSoFar);
    }
    return maxSoFar;
}

int main(){
    vector<int> array = {1,2,-1,3,5,-9,-2};
    int max_num = maxSumSubarray(array);
    cout<<"Maximum sum obtained from sub array - "<<max_num <<endl;
    return 0;
}

File: dp/kadanes/python/max_sum_subarray.py
def max_sum_subarray(array):
    max_ending_here = array[0]
    max_so_far = array[0]
    for num in array[1:]:
        max_ending_here = max(max_ending_here+num, num)
        max_so_far = max(max_ending_here, max_so_far)
    return max_so_far


if __name__=='__main__':
    array = [1,2,-1,3,5,-9,-2]
    max_num = max_sum_subarray(array)
    print(max_num)